### PR TITLE
8252794: Creation of JNIMethodBlock should be done with a leaf lock

### PR DESCRIPTION
**Pull request by robehn**
JVM TI uses handshakes and sometimes need jmethod ids.
The CLD have an area where these ids are saved, which is created on first request for an id.

The creation of this area should be done with a leaf lock.
In W.I.P. asynchronous handshake change-set we use a mutex instead of a semaphore for serializing the handshakes,
we thus get into lock ranks asserts.

The CLD metaspace lock cannot be pushed down but we can use the same lock as for creating jmethod ids: JmethodIdCreation_lock.
And push that down. (special-2 is picked to be compatible with W.I.P. asynchronous handshake change-set)

The reason for CLD metaspace lock being used now is that we have one per CLD instead of one global one.

Testing passes: t1-3

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8252794](https://bugs.openjdk.java.net/browse/JDK-8252794): Creation of JNIMethodBlock should be done with a leaf lock

### Reviewers
 * [Coleen Phillimore](https://openjdk.java.net/census#coleenp) (@coleenp - **Reviewer**) ⚠️ Review applies to ec4eb407102bbf7b842bd6888765f2e5e34f8e69
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/53/head:pull/53`
`$ git checkout pull/53`

### CODE REVIEW

**Review by coleenp — APPROVED** [review]

File: src/hotspot/share/oops/method.cpp [hunk 1]
```diff
@@ -2187,7 +2187,7 @@ void Method::ensure_jmethod_ids(ClassLoaderData* loader_data, int capacity) {
     // Have to add jmethod_ids() to class loader data thread-safely.
     // Also have to add the method to the list safely, which the cld lock
     // protects as well.
-    MutexLocker ml(cld->metaspace_lock(),  Mutex::_no_safepoint_check_flag);
+    MutexLocker ml(JmethodIdCreation_lock,  Mutex::_no_safepoint_check_flag);
```
Comment (dholmes-ora) [inline src/hotspot/share/oops/method.cpp:2190]: The comment above this states:
 // Also have to add the method to the list safely, which the cld lock
 // protects as well.
I don't know what list is being referred to here but either the comment is incorrect or else the change of lock is incorrect!

File: src/hotspot/share/oops/method.cpp [hunk 2]
```diff
@@ -2211,7 +2211,7 @@ jmethodID Method::make_jmethod_id(ClassLoaderData* loader_data, Method* m) {
...
```
Comment (dholmes-ora) [inline src/hotspot/share/oops/method.cpp:2214]: Ditto

**Review by dholmes-ora — CHANGES_REQUESTED** [review]
The locking strategy in this code is somewhat unclear.

File: src/hotspot/share/oops/method.cpp (hunk 1 above)
Comment (robehn) [reply src/hotspot/share/oops/method.cpp:2190]: We create the list with:
`cld->set_jmethod_ids(new JNIMethodBlock(capacity));`
And add to the list:
`return (jmethodID)cld->jmethod_ids()->add_method(m);`

I removed 'cld' from comment.

File: src/hotspot/share/oops/method.cpp (hunk 2 above)
Comment (robehn) [reply src/hotspot/share/oops/method.cpp:2214]: Ditto fix

**Review by dholmes-ora — APPROVED** [review]
Update is good.

**Comment by robehn** [comment]
> Update is good.

Thank you. As the update was trivial I'll go head and integrate.

**Pull request closed without merging**